let 제거 과제에 대한 PR 요청합니다.

Ubermensch0608 opened the pull request:
안녕하세요 트레이너님!

리뷰 요청드립니다.

let을 사용하지 않으면서 state변경하는 또 다른 방법으로 class 를 사용하는 것을 생각했는데
결국 class 내에서도 객체를 사용하여 관리하는 것과 비슷하다고 생각되어 우선 객체에서 값이 변경되는 것으로 구현했습니다.
class를 이용하여 state를 관리하는 방식이 class based component 에서 사용하는 것으로 알고 있어서 이 쪽이 더 괜찮은 관리 방식일까요?

Ubermensch0608 committed:
Feat: state를 조작하는 버튼구현

버튼을 누르면 값이 1씩 증가하는 기능을 구현했습니다. 특정 값으로 변경도 가능합니다.

Ubermensch0608 committed:
Update: let을 사용하지 않는 방식으로 state를 업데이트하도록 변경

let을 사용할 경우 변수가 재할당 되므로 state라는 객체를 만든 뒤 그 안에서 여러 값을 다룰 수 있도록 변경하였습니다.

state에서 count 외 여러 값을 사용할 수 있어 확장성이 높을 것이라 생각됩니다.

Ubermensch0608 committed:
Update: 테스트 통과를 위해 세부 값 변경

버튼 내부 텍스트를 '클릭!' 에서 'Click me!'로 변경하였습니다.
버튼 내부 count 값을 감싸는 ()를 추가하였습니다.

Ubermensch0608 committed:
Chore: index.js 내 불필요한 코드 삭제

eslint 관련 중복 코드 삭제했습니다.

hannut91 commented:
객체를 선언했을 때는, 객체 자체를 재할당 할 수는 없지만 객체에 속성들은 재할당이 가능하다는 것을 응용해서 문제를 해결해 주셨네요!
아쉽게도 여전히 어떤 값이 변경된다는 사실에는 변함이 없습니다. 과제의 의도는 변수의 사용 대신 상수를 사용해서 문제를 해결하는 것인데요. 변수가 변한다는 사실 자체만으로도 프로그램의 복잡성을 많이 증가시킵니다. 어떤 값을 내가 사용할 때 이 값이 현재 어떤 값인지 예측하기 어렵기 때문이죠

hannut91 commented on .eslintrc.js:
```diff
@@ -1,33 +1,29 @@
+        },
+    },
+    rules: {},
+};
```
이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

프리티어 혹은 eslint에서 이러한 것을 자동으로 할 수 있는지 확인해보는 것도 좋겠습니다

Ubermensch0608 replied:
prettier에 자동으로 마지막 빈 줄 하나 추가해주는 설정하여 해당 오류 해결했습니다!
이전부터 이러한 설정이 왜 있는지 궁금했었는데 git 관련해서 발생할 수 있는 문제도 방지할 수 있네요. 감사합니다 :)

hannut91 commented on src/index.js:
```diff
@@ -1,2 +1,62 @@
+        render();
+    }
+
+    const element = ( <
+        div >
+        <
+        button type = "button"
+        onClick = { clickHandler } >
+        Click me!({ state.count }) <
+        /button> <
+        p > {
+            [1, 2, 3].map(i => ( <
+                button type = "button"
+                onClick = {
+                    () => clickNumberHandler(i) } >
+                ({ i }) <
+                /button>
+            ))
+        } <
+        /p> <
+        /div>
+    );
```
여기는 뭔가 이상하게 되었네요.

Ubermensch0608 replied:
vscode 확장기능 중에 위 문제를  발생시키는 것이 있어 제거 후에 해결했습니다!

Ubermensch0608 committed:
Refactor: 변수를 재할당 시키지 않고 업데이트된 count값 렌더링 구현

액션이 발생할 때 render 함수의 argument 요소 값을 참조하여 원하는 업데이트가 일어나도록 수정하였습니다.

다음은 수정이 일어난 이유입니다.
1. 이전에 구현했던 상수 객체를 두고 객체 내부 값을 변경시켜 구현했지만 이 역시 값의 재할당입니다.
2. 값의 상태를 예측하기 어려웠습니다.

Ubermensch0608 commented:
> 객체를 선언했을 때는, 객체 자체를 재할당 할 수는 없지만 객체에 속성들은 재할당이 가능하다는 것을 응용해서 문제를 해결해 주셨네요!
> 
> 아쉽게도 여전히 어떤 값이 변경된다는 사실에는 변함이 없습니다. 과제의 의도는 변수의 사용 대신 상수를 사용해서 문제를 해결하는 것인데요. 변수가 변한다는 사실 자체만으로도 프로그램의 복잡성을 많이 증가시킵니다. 어떤 값을 내가 사용할 때 이 값이 현재 어떤 값인지 예측하기 어렵기 때문이죠

피드백 감사합니다!
말씀해주신 '예측이 어려운 값'이라는 것에 초점을 맞추어 보안해봤습니다.

Ubermensch0608 committed:
Chore: github ci end of line 오류 수정

prettier 및 eslint 설정 수정하여 위 오류 해결하였습니다.

daadaadaah commented on public/index.html:
```diff
@@ -1,11 +1,14 @@
 <!DOCTYPE html>
 <html>
-  <head>
+
```
앞으로 PR 날리실 때, 이런 불필요한 줄은 정리해서 올려주시는게 좋아요!ㅎㅎ

Ubermensch0608 replied:
넵 PR 전에 불필요한 코드 신경써보겠습니다!

Ubermensch0608 committed:
Chore: github ci 오류 수정

Ubermensch0608 committed:
Chore: 불필요한 줄 내림 제거

Ubermensch0608 committed:
Chore: index.html 마지막 줄 추가

github ci 오류 해결을 위해 마지막 줄 추가했습니다.
지금까지 prettier 설정으로 .js 파일에는 마지막 라인 추가가 되었는데 .html은 안 되고 있는지 몰랐네요. 해당 설정 추가하였습니다.

Ubermensch0608 committed:
Chore: 코드 정리
